### PR TITLE
Fixed #25916 -- Added lastmod support to sitemaps.views.index

### PR DESCRIPTION
**Pull request by mgd020**
Allowed sitemap_index.xml to supply lastmod values for each sitemap.

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
Please take a look at our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist). From a quick glance, at least a mention in the release notes and a versionadded annotation are missing. Thanks!

**Comment by mgd020** [comment]
Done.

File: docs/ref/contrib/sitemaps.txt [hunk 1]
```diff
@@ -249,6 +253,17 @@ Sitemap class reference
         be generated using all of your :setting:`LANGUAGES`. The default is
         ``False``.
 
+    .. function:: Sitemap.get_latest_lastmod()
+
+        **Optional.** A method that returns the latest value returned by :attr:`~Sitemap.lastmod()`.
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: Please wrap at 79 characters.

File: docs/ref/contrib/sitemaps.txt [hunk 2]
```diff
@@ -249,6 +253,17 @@ Sitemap class reference
         be generated using all of your :setting:`LANGUAGES`. The default is
         ``False``.
 
+    .. function:: Sitemap.get_latest_lastmod()
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: function -> method

File: docs/ref/contrib/sitemaps.txt [hunk 3]
```diff
@@ -188,6 +188,10 @@ Sitemap class reference
         Django respond appropriately to requests with an ``If-Modified-Since``
         header which will prevent sending the sitemap if it hasn't changed.
 
+        .. versionadded:: 1.10
+
+            Default template now includes time value of :attr:`~Sitemap.lastmod`
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: "The value of :attr:`~Sitemap.lastmod` was added to the default sitemap template."

File: docs/ref/contrib/sitemaps.txt [hunk 4]
```diff
@@ -188,6 +188,10 @@ Sitemap class reference
         Django respond appropriately to requests with an ``If-Modified-Since``
         header which will prevent sending the sitemap if it hasn't changed.
 
+        .. versionadded:: 1.10
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: I think versionchanged would be more appropriate here.

File: docs/ref/contrib/sitemaps.txt [hunk 5]
```diff
@@ -249,6 +253,17 @@ Sitemap class reference
...
+    .. function:: Sitemap.get_latest_lastmod()
+
+        **Optional.** A method that returns the latest value returned by :attr:`~Sitemap.lastmod()`.
+
+        By default :meth:`~Sitemap.get_latest_lastmod()` returns:
+
+        * If :attr:`~Sitemap.lastmod` is an attribute: :attr:`~Sitemap.lastmod`.
+        * If :attr:`~Sitemap.lastmod` is a method: ``None``.
+
+        .. versionadded:: 1.10
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: we usually put this at the top of the section (above "Optional")

File: django/contrib/sitemaps/__init__.py [hunk 6]
```diff
@@ -150,5 +153,10 @@ def lastmod(self, item):
             return getattr(item, self.date_field)
         return None
 
+    def get_latest_lastmod(self):
```
Comment (timgraham) [inline django/contrib/sitemaps/__init__.py]: I think this method is missing tests -- none fail if I comment it out.

File: django/contrib/sitemaps/__init__.py [hunk 7]
```diff
@@ -67,6 +67,9 @@ def items(self):
     def location(self, obj):
         return obj.get_absolute_url()
 
+    def get_latest_lastmod(self):
```
Comment (timgraham) [inline django/contrib/sitemaps/__init__.py]: Might be useful to add some unittests for these methods. I don't think we have any tests that don't go through the full request/response cycle, but it might make testing a bit easier to follow if that's not required for every test.

File: django/contrib/sitemaps/__init__.py [hunk 8]
```diff
@@ -67,6 +67,9 @@ def items(self):
...
+        return self.lastmod if hasattr(self, 'lastmod') and not callable(self.lastmod) else None
```
Comment (timgraham) [inline django/contrib/sitemaps/__init__.py]: No tests fail if I remove `and not callable(self.lastmod)`

File: docs/releases/1.10.txt [hunk 9]
```diff
@@ -116,7 +116,10 @@ Minor features
 :mod:`django.contrib.sitemaps`
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-* ...
+* Added support for generated ``<sitemapindex>`` to include ``<lastmod>`` 
+  through  :meth:`~django.contrib.sitemaps.Sitemap.get_latest_lastmod()`.
+
+* Updated generated ``<sitemap>`` to include the time value of ``<lastmod>``.
```
Comment (timgraham) [inline docs/releases/1.10.txt]: "generated" -> "the default sitemap template"?

File: docs/ref/contrib/sitemaps.txt (hunk 5 above)
Comment (mgd020) [inline docs/ref/contrib/sitemaps.txt]: put what exactly, where exactly?

I copied the format/layout of `attribute:: Sitemap.priority` above (line 216).

Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: The `.. versionadded:: 1.10` annotation should be above "Optional".

File: django/contrib/sitemaps/__init__.py [hunk 10]
```diff
@@ -122,7 +136,7 @@ def _urls(self, page, protocol, domain):
             url_info = {
                 'item': item,
                 'location': loc,
-                'lastmod': lastmod,
+                'lastmod': sitemap_time(lastmod),
```
Comment (timgraham) [inline django/contrib/sitemaps/__init__.py]: Could this type change be backwards incompatible for custom sitemap templates?

File: django/contrib/sitemaps/__init__.py [hunk 11]
```diff
@@ -44,9 +46,17 @@ def ping_google(sitemap_url=None, ping_url=PING_URL):
     urlopen("%s?%s" % (ping_url, params))
 
 
+def sitemap_time(t):
+    if t is None:
+        return None
+    if isinstance(t, datetime.datetime) and (t.tzinfo is None or t.tzinfo.utcoffset(t) is None):
```
Comment (timgraham) [inline django/contrib/sitemaps/__init__.py]: Are you sure this isn't duplicating a function in `django.utils.timezone`? Could you please add a docstring to the function?

File: django/contrib/sitemaps/views.py [hunk 12]
```diff
@@ -34,9 +35,10 @@ def index(request, sitemaps,
         protocol = req_protocol if site.protocol is None else site.protocol
         sitemap_url = reverse(sitemap_url_name, kwargs={'section': section})
         absolute_url = '%s://%s%s' % (protocol, req_site.domain, sitemap_url)
-        sites.append(absolute_url)
+        lastmod = sitemap_time(site.get_latest_lastmod())
+        sites.append({'location': absolute_url, 'lastmod': lastmod})
         for page in range(2, site.paginator.num_pages + 1):
-            sites.append('%s?p=%s' % (absolute_url, page))
+            sites.append({'location': '%s?p=%s' % (absolute_url, page), 'lastmod': lastmod})
```
Comment (timgraham) [inline django/contrib/sitemaps/views.py]: Again, isn't this backwards-incompatible with custom templates?

File: docs/ref/contrib/sitemaps.txt [hunk 13]
```diff
@@ -173,13 +173,13 @@ Note:
 
         **Optional.** Either a method or attribute.
 
-        If it's a method, it should take one argument -- an object as returned by
-        :attr:`~Sitemap.items()` -- and return that object's last-modified date/time, as a Python
-        ``datetime.datetime`` object.
+        If it's a method, it should take one argument -- an object as returned
```
Comment (timgraham) [inline docs/ref/contrib/sitemaps.txt]: I merged these line wrapping changes in a496d10a8c3b8b3080f0daa7348b33aed790a967 since it's better not to include unrelated cleanups in a feature like this.

**Comment by timgraham** [comment]
Please rebase your branch and squash commits when updating. Also, please uncheck "Patch needs improvement" on the ticket so that it returns to the review queue.

**Comment by timgraham** [comment]
Closing due to inactivity. Please send a new PR if you return to it.
